### PR TITLE
S3Bucket and Bucket

### LINKED ISSUE #109: Buckets: APIs, controllers, and provisioning
Similar to the support we have for MySQL, support for object storage Buckets should be implemented as well.  This includes:

* concrete provider specific types/CRDs to represent a bucket with each cloud provider
* concrete controllers to manage bucket lifecycles at users request
* abstract bucket CRD to represent the general bucket common across all providers
* abstract bucket controller to handle abstract bucket events, call the correct concrete provisioner to instantiate a concrete bucket, and bind the 2 together

### PR DESCRIPTION
**Pull request by lukeweber**
Fixes #109

### CODE REVIEW

File: pkg/apis/aws/storage/v1alpha1/bucket_types.go [hunk 1]
```diff
@@ -70,6 +127,91 @@ type S3BucketList struct {
...
+		return b.Status.ConnectionSecretRef.Name
+	} else if b.Spec.ConnectionSecretNameOverride != "" {
+		return b.Spec.ConnectionSecretNameOverride
+	}
+	return b.Name
+}
+
+// OwnerReference to use this instance as an owner
+func (b *S3Bucket) OwnerReference() metav1.OwnerReference {
+	if b.APIVersion == "" {
```
[the remark this answers is not in the document]
Comment (lukeweber) [reply pkg/apis/aws/storage/v1alpha1/bucket_types.go]: One thought, this is all copied code. Would be nice if we had a "reference" file or found a way for this code to be shared.

File: pkg/clients/aws/iam/iam.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	"github.com/aws/aws-sdk-go-v2/service/iam/iamiface"
+)
+
+const (
+	policyArn = "arn:aws:iam::%s:policy/%s"
+)
+
+// Service defines IAM Client operations
+type Service interface {
```
[the remark this answers is not in the document]
Comment (lukeweber) [reply pkg/clients/aws/iam/iam.go]: Happy to rename. On a downside, you get iam.IamClient which stutters.

[the remark this answers is not in the document]
Comment (lukeweber) [reply pkg/clients/aws/iam/iam.go]: Also I left IAM client separate on the basis that maybe it gets broken out in the near future or used in other components depending on what you want to do.

File: pkg/clients/aws/s3/s3.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+	policy := iamc.PolicyDocument{
+		Version:   "2012-10-17",
+		Statement: []iamc.StatementEntry{},
+	}
+
+	if spec.LocalPermission != nil {
+		if *spec.LocalPermission == storage.ReadOnlyPermission {
+			policy.Statement = append(policy.Statement, read)
+		} else if *spec.LocalPermission == storage.WriteOnlyPermission {
+			policy.Statement = append(policy.Statement, read)
```
[the remark this answers is not in the document]
Comment (lukeweber) [reply pkg/clients/aws/s3/s3.go]: good catch. thanks

File: pkg/clients/aws/iam/iam.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+	CreateUser(username *string) (*iam.AccessKey, error)
+	DeleteUser(username *string) error
+	CreatePolicyAndAttach(username *string, policyName *string, policyDocument *string) (*string, error)
+	GetPolicyVersion(policyName *string) (*string, error)
+	UpdatePolicy(policyName *string, policyDocument *string) (*string, error)
+	DeletePolicyAndDetach(username *string, policyName *string) error
```
[the remark this answers is not in the document]
Comment (lukeweber) [reply pkg/clients/aws/iam/iam.go]: Agreed. Pointers have been useful though in a few instances in CRDs to know the difference between not set and a desire to set to empty just as a note.

File: pkg/controller/storage/bucket/bucket.go [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+		return result, err
+	}
+
+	// Check for deletion
+	if bucket.DeletionTimestamp != nil && bucket.Status.Condition(corev1alpha1.Deleting) == nil {
+		return r.delete(bucket)
+	}
+
+	// Add finalizer
+	if !util.HasFinalizer(&bucket.ObjectMeta, finalizer) {
```
[the remark this answers is not in the document]
Comment (lukeweber) [reply pkg/controller/storage/bucket/bucket.go]: This is also just copied. In my mind this should be a generic resource handler logic if possible. Would love to help solve this as we're only going to have more and should speed up reviews.

https://github.com/crossplaneio/crossplane/blob/a70ff14dd3edd4035e7bc3b95f183dd3db5a9ac6/pkg/controller/storage/mysql/mysqlinstance.go#L301-L306

[the remark this answers is not in the document]
Comment (jbw976) [reply pkg/controller/storage/bucket/bucket.go]: @lukeweber @ichekrygin this looks exactly like the duplication I set up to remove/avoid with the generic resource controller i wrote for the addition of postgres support. See https://github.com/jbw976/crossplane/blob/sql-controller/pkg/controller/core/resource_controller.go#L97

And a concrete resource controller making use of it: https://github.com/jbw976/crossplane/blob/sql-controller/pkg/controller/storage/sql/postgresinstance.go#L87

I think after that's in master, we should try to convert other controllers (such as this bucket controller) to use the generic resource controller and eliminate a lot of this duplicated code.

**Comment by lukeweber** [comment]
@ichekrygin  - thanks for a thorough review. Copied code seemed to be a stumbling point for me on integration, but I made all the updates.

**Pull request merged**